the-collab-lab/tcl-19-smart-shopping-list
# Connects  app to firestore cloud storage

## Description

In this pull request(PR), we:
1. Added [firebase](https://www.npmjs.com/package/firebase) and [react-firebase-hooks](https://www.npmjs.com/package/react-firebase-hooks) as project dependencies.
2.  Created `components` and  `styles` directories within `src` directory. This provides structure to our code and makes it easy to read, search and maintain it.
3. Added code which connects the app to firestore cloud storage. 

**NOTE**
In the process of preparing this PR, we came across the following which we believe are worth mentioning here. Others reviewing this PR might find them useful.
- According to [firebase docs](https://firebase.google.com/docs/web/setup#add_firebase_to_your_app) and [this stackoveflow question](https://stackoverflow.com/questions/37482366/is-it-safe-to-expose-firebase-apikey-to-the-public), it is not a security risk to expose your [firestore configuration details](https://github.com/the-collab-lab/tcl-19-smart-shopping-list/blob/main/src/lib/firebase.js) though it includes your `API` key. It could however be a security risk  to store sensitive credentials like API keys for other platforms in a  gitHub repository. The best place for them is in the `.env` file which you then ignore in `.gitignore`. 
- You can write react components with `.js` or `.jsx` extensions. Though not commonly used and we haven't found a reason why it would be preferable, the latter is compiled by babel to `.js`.
- It is important to have standard naming conventions/rules when working in a team so that each person follows the same convention/rules when adding changes. 


## Related Issue

closes #1 


## Acceptance Criteria



## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|   ✓ | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before

We had components which came with Create React App and firebase configuration files in the `lib` directory.


![image](https://user-images.githubusercontent.com/52580190/103740663-bc0f1e80-5008-11eb-97d8-11f93863a00c.png)


### After

We added a `form` which one can use to add items to the shopping list and a component for displaying the list of items in the database.

![image](https://user-images.githubusercontent.com/52580190/103741288-d4cc0400-5009-11eb-827a-3001c8d42266.png)





## Testing Steps / QA Criteria

To see the features we added, you can follow the following steps:
1. Pull the changes in the remote branch to your local setup using `npm pull origin mj-ju-connect-to-firestore` after checking it out. Assuming your `origin` has been configured to point to `https://github.com/the-collab-lab/tcl-19-smart-shopping-list.git`.
2. Run `npm run start`. React will launch the development server on port 3000. If  something else is running on port 3000, you will be prompted to decide whether to run it on another port. You can navigate to `http://localhost:3000` or `http://localhost:PORT` if the server has been launched on a different `PORT`.
3. You can fill the form and submit. The item you have submitted will be added to the database and then displayed in the browser.
4. You can also add items from your firestore dashboard. Under `items` collection, you can click `Add document` and add an item. After which you should be able to see the added item in the browser. We are using the field names `item` and `how_much`.
